### DISCUSSION TITLE
[Security] How to prevent XSS in all _rsc calls

### DISCUSSION THREAD

**Discussion by motss** [vercel/next.js#66560] [Help]
## Background

XSS is actually possible in a `GET /?_rsc` API call in Next.js such as `GET /https://example.com/?_rsc=abc&<img src= onerror=alert('test')>` where malicious script can be injected from the URL and the response will contain the injected script.

```js
0:["<some_id>",[["children","__PAGE__?{\\"<img src\\":\\" onerror=alert('test')>\\"}",["__PAGE__?{\\"<img src\\":\\" onerror=alert('test')>\\"}",{}],null,null]]]\
```

I wonder how the Next.js community protects these `_rsc` API calls from XSS. Could you please shed some light on this? Thanks.

### Additional information

```js
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.5.0: Wed May  1 20:13:18 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6030
Binaries:
  Node: 20.10.0
  npm: 10.2.3
  Yarn: N/A
  pnpm: 9.1.2
Relevant Packages:
  next: 14.1.2
  eslint-config-next: 13.4.19
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: standalone
```

### Example

GET /https://example.com/?_rsc=abc&<img src= onerror=alert('test')>

**Comment by icyJoseph** [comment]
Hi, could you show us a video of the alert triggering? I tried a couple of tricks I knew, but I can't get it to actually XSS, everything I try gets escaped.
